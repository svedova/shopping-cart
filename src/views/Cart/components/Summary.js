import React, { PureComponent, Fragment } from "react";
import PropTypes from "prop-types";
import Styled from "styled-components";
import { connect } from "react-redux";
import { props } from "@/ui";
import { removeFromCart } from "../actions/cart";
import { getTotal, getSubtotal } from "@/helpers/functions";

const StyledSummary = Styled.div`
  position: absolute;
  top: 100%;
  right: 0;
  width: 350px;
  background: black;
  padding: ${props.paddingM};
  box-shadow: ${props.shadowM};
  text-align: left;
  z-index: 999;
`;

const Row = Styled.div`
  display: flex;
  justify-content: space-between;
  margin-bottom: ${props.marginL};
  border-top: ${p => (p.withBorderTop ? `1px solid ${props.gray50}` : "")};
  padding-top: ${p => (p.withBorderTop ? props.paddingL : 0)};
  
  &:last-child {
    margin-bottom: 0;
  }
`;

const Col = Styled.div`
  flex: 1 1 auto;
  
  &:first-child {
    width: 69.6%;
  }
  
  &:last-child {
    max-width: 33px;
    align-self: flex-end;
    text-align: right;
  }
`;

const Quantity = Styled.span`
  display: inline-block;
  color: ${props.gray50};
  font-size: 0.8rem;
  margin-right: ${props.marginM};
  width: 30px;
`;

const Remove = Styled.div`
  color: red;
  user-select: none;
  cursor: pointer;
`;

const ToCheckout = Styled.div`
  border-top: 1px solid ${props.gray50};
  text-align: center;
  padding: ${props.paddingM};
  
  a {
    color: white;
    text-decoration: none;
    
    &:hover {
      color: orange;
    }
  }
`;

export class Summary extends PureComponent {
  static defaultProps = {
    items: {},
    cart: {
      quantities: {},
      ids: []
    }
  };

  static propTypes = {
    cart: PropTypes.object,
    items: PropTypes.array,
    removeFromCart: PropTypes.func,
    className: PropTypes.any,
    action: PropTypes.node
  };

  /**
   * Return brand, title if there is a brand, otherwise only the title.
   *
   * @param item
   * @return {string}
   */
  normalizedName = item => {
    const { cart } = this.props;
    const name = [item.brand, item.title].filter(i => i).join(", ");
    const quantity = cart.quantities[item.id];
    return (
      <Fragment>
        <Quantity>(x{quantity})</Quantity>
        {name}
      </Fragment>
    );
  };

  price = item => {
    const { cart } = this.props;
    const price = (item.discounted || item.price) * cart.quantities[item.id];
    return `CHF ${price.toFixed(2)}`;
  };

  getTotal = () => {
    const { cart, items } = this.props;
    return getTotal({ items, cart: cart.quantities });
  };

  getSubtotal = () => {
    const { cart, items } = this.props;
    return getSubtotal({ items, cart: cart.quantities });
  };

  render() {
    const { cart, items, removeFromCart, action } = this.props;
    const itemsInCart = cart.ids.map(id => items.filter(i => i.id === id)[0]);
    const total = this.getTotal();
    const subtotal = this.getSubtotal();

    return (
      <StyledSummary className={this.props.className}>
        {itemsInCart.map(item => (
          <Row key={item.id}>
            <Col>{this.normalizedName(item)}</Col>
            <Col>{this.price(item)}</Col>
            <Col onClick={() => removeFromCart(item)}>
              <Remove className="fas fa-minus-circle" />
            </Col>
          </Row>
        ))}
        <Row withBorderTop>Subtotal: CHF {subtotal}</Row>
        <Row withBorderTop>Discount: CHF {(+subtotal - +total).toFixed(2)}</Row>
        <Row withBorderTop>Total: CHF {total}</Row>
        <ToCheckout>{action}</ToCheckout>
      </StyledSummary>
    );
  }
}

export const mapStateToProps = state => ({
  cart: state.cart,
  items: state.items.all
});

export default connect(mapStateToProps, { removeFromCart })(Summary);
