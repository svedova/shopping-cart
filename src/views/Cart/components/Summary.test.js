import React from "react";
import { mount } from "enzyme";
import { Summary, mapStateToProps } from "./Summary";
import mocks from "@/tests/mocks";

describe("(Component) Summary", () => {
  let comp;

  beforeEach(() => {
    comp = mount(<Summary items={mocks.items.all} cart={mocks.cart} />);
  });

  test("should mount", () => {
    expect(comp.html()).toMatchSnapshot();
  });

  test("should match the snapshot for mapStateToProps", () => {
    expect(
      mapStateToProps({ cart: mocks.cart, items: mocks.items })
    ).toMatchSnapshot();
  });

  test("should return the total correctly", () => {
    const inst = comp.instance();
    expect(inst.getTotal()).toBe("16.86");
    expect(inst.getSubtotal()).toBe("17.10");
  });

  test("should return the correct price", () => {
    const inst = comp.instance();
    expect(inst.price(mocks.items.all[0])).toBe("CHF 0.75");
    expect(mocks.items.all[0].price).toBe(0.25);
    expect(mocks.cart.quantities[mocks.items.all[0].id]).toBe(3);
  });

  test("should return the normalized name", () => {
    const inst = comp.instance();
    const item1 = mocks.items.all[0];
    const item2 = mocks.items.all[2];
    const div1 = mount(<div>{inst.normalizedName(item1)}</div>);
    const div2 = mount(<div>{inst.normalizedName(item2)}</div>);

    expect(div1.text()).toBe("(x3)Apple");
    expect(div2.text()).toBe("(x4)Barilla, Penne");
  });
});
