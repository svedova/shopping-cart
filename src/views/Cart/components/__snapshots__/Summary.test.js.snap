// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Summary should match the snapshot for mapStateToProps 1`] = `
Object {
  "cart": Object {
    "ids": Array [
      "7sRCjMPmSX3XNgzn074v",
      "7fovkddLybsCiOoV5jf7",
      "LdKHLDJpnbZR04VuPxE5",
      "Ft5wrWvYf0UrQf0G6Ahg",
      "fzjeAkCkmNQHpPfZ5fdJ",
      "uy4PC76zFiL808V2OdgQ",
      "G1D8XDDwIChYUabzoUmE",
    ],
    "quantities": Object {
      "7fovkddLybsCiOoV5jf7": 3,
      "7sRCjMPmSX3XNgzn074v": 3,
      "Ft5wrWvYf0UrQf0G6Ahg": 4,
      "G1D8XDDwIChYUabzoUmE": 1,
      "LdKHLDJpnbZR04VuPxE5": 3,
      "fzjeAkCkmNQHpPfZ5fdJ": 3,
      "uy4PC76zFiL808V2OdgQ": 2,
    },
  },
  "items": Array [
    Object {
      "category": "fruit",
      "discount": 0,
      "id": "7fovkddLybsCiOoV5jf7",
      "price": 0.25,
      "title": "Apple",
    },
    Object {
      "category": "fruit",
      "discount": 15,
      "discounted": 0.43,
      "id": "7sRCjMPmSX3XNgzn074v",
      "price": 0.5,
      "title": "Papaya",
    },
    Object {
      "brand": "Barilla",
      "category": "dried pasta",
      "id": "Ft5wrWvYf0UrQf0G6Ahg",
      "price": 2,
      "title": "Penne",
    },
    Object {
      "brand": "Mister Rice",
      "category": "rice",
      "id": "G1D8XDDwIChYUabzoUmE",
      "price": 1.5,
      "title": "Jasmin",
    },
    Object {
      "category": "fruit",
      "id": "LdKHLDJpnbZR04VuPxE5",
      "price": 0.3,
      "title": "Orange",
    },
    Object {
      "category": "fruit",
      "discount": 10,
      "discounted": 0.14,
      "id": "fzjeAkCkmNQHpPfZ5fdJ",
      "price": 0.15,
      "title": "Banana",
    },
    Object {
      "brand": "Tesco",
      "category": "dried pasta",
      "discount": 5,
      "discounted": 1.66,
      "id": "qvjW0xRluPKogdwYAXzf",
      "price": "1.75",
      "title": "Spaghetti",
    },
    Object {
      "brand": "Barilla",
      "category": "dried pasta",
      "id": "uy4PC76zFiL808V2OdgQ",
      "price": 2,
      "title": "Spaghetti",
    },
  ],
}
`;

exports[`(Component) Summary should mount 1`] = `"<div class=\\"sc-htoDjs ciqLhv\\"><div class=\\"sc-dnqmqq hbtSjX\\"><div class=\\"sc-iwsKbI jkme\\"><span class=\\"sc-gZMcBi geLwpI\\">(x3)</span>Papaya</div><div class=\\"sc-iwsKbI jkme\\">CHF 1.29</div><div class=\\"sc-iwsKbI jkme\\"><div class=\\"fas fa-minus-circle sc-gqjmRU kwWzOp\\"></div></div></div><div class=\\"sc-dnqmqq hbtSjX\\"><div class=\\"sc-iwsKbI jkme\\"><span class=\\"sc-gZMcBi geLwpI\\">(x3)</span>Apple</div><div class=\\"sc-iwsKbI jkme\\">CHF 0.75</div><div class=\\"sc-iwsKbI jkme\\"><div class=\\"fas fa-minus-circle sc-gqjmRU kwWzOp\\"></div></div></div><div class=\\"sc-dnqmqq hbtSjX\\"><div class=\\"sc-iwsKbI jkme\\"><span class=\\"sc-gZMcBi geLwpI\\">(x3)</span>Orange</div><div class=\\"sc-iwsKbI jkme\\">CHF 0.90</div><div class=\\"sc-iwsKbI jkme\\"><div class=\\"fas fa-minus-circle sc-gqjmRU kwWzOp\\"></div></div></div><div class=\\"sc-dnqmqq hbtSjX\\"><div class=\\"sc-iwsKbI jkme\\"><span class=\\"sc-gZMcBi geLwpI\\">(x4)</span>Barilla, Penne</div><div class=\\"sc-iwsKbI jkme\\">CHF 8.00</div><div class=\\"sc-iwsKbI jkme\\"><div class=\\"fas fa-minus-circle sc-gqjmRU kwWzOp\\"></div></div></div><div class=\\"sc-dnqmqq hbtSjX\\"><div class=\\"sc-iwsKbI jkme\\"><span class=\\"sc-gZMcBi geLwpI\\">(x3)</span>Banana</div><div class=\\"sc-iwsKbI jkme\\">CHF 0.42</div><div class=\\"sc-iwsKbI jkme\\"><div class=\\"fas fa-minus-circle sc-gqjmRU kwWzOp\\"></div></div></div><div class=\\"sc-dnqmqq hbtSjX\\"><div class=\\"sc-iwsKbI jkme\\"><span class=\\"sc-gZMcBi geLwpI\\">(x2)</span>Barilla, Spaghetti</div><div class=\\"sc-iwsKbI jkme\\">CHF 4.00</div><div class=\\"sc-iwsKbI jkme\\"><div class=\\"fas fa-minus-circle sc-gqjmRU kwWzOp\\"></div></div></div><div class=\\"sc-dnqmqq hbtSjX\\"><div class=\\"sc-iwsKbI jkme\\"><span class=\\"sc-gZMcBi geLwpI\\">(x1)</span>Mister Rice, Jasmin</div><div class=\\"sc-iwsKbI jkme\\">CHF 1.50</div><div class=\\"sc-iwsKbI jkme\\"><div class=\\"fas fa-minus-circle sc-gqjmRU kwWzOp\\"></div></div></div><div class=\\"sc-dnqmqq fOyjxt\\">Subtotal: CHF 17.10</div><div class=\\"sc-dnqmqq fOyjxt\\">Discount: CHF 0.24</div><div class=\\"sc-dnqmqq fOyjxt\\">Total: CHF 16.86</div><div class=\\"sc-VigVT fbBmcQ\\"></div></div>"`;
